src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -229,43 +231,136 @@ private void ReconnectWorkersIfNeeded(List<string> shardsToReconnect)
+                    {
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+                        (BlittableJsonReaderObject metadata, string id, string changeVector) = BatchFromServer.GetMetadataFromBlittable(doc.Data);
+                        lastReceivedChangeVector = changeVector;
+                        SubscriptionConnection.WriteDocumentOrException(context, writer, document: null, doc.Data, metadata, doc.Exception, id, null, null, null);
+                        docsToFlush++;
+
+                        if (await SubscriptionConnection.FlushBatchIfNeededAsync(sendingCurrentBatchStopwatch, SubscriptionId, writer, buffer, TcpConnection, stats: null, _logger, docsToFlush, CancellationTokenSource.Token))
+                            continue;
+
+                        docsToFlush = 0;
+                        sendingCurrentBatchStopwatch.Restart();
+                    }
+                    batch.LastSentChangeVectorInBatch = lastReceivedChangeVector;
+                    AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Flushing docs collected from shard '{batch.ShardNumber}'"));
                 }
-                catch (Exception e)
+
+                DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Egor, DevelopmentHelper.Severity.Major, "https://issues.hibernatingrhinos.com/issue/RavenDB-16279");
+                SubscriptionConnection.WriteEndOfBatch(writer);
+
+                AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Flushing docs collected from '{shards}'"));
+                await SubscriptionConnection.FlushDocsToClientAsync(SubscriptionId, writer, buffer, TcpConnection, stats: null, _logger, docsToFlush, endOfBatch: true, CancellationTokenSource.Token);
+            }
+
+            await WaitForClientAck(replyFromClientTask);
+            AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, $"Got ack for shards '{shards}' from client '{TcpConnection.TcpClient.Client.RemoteEndPoint}'."));
+        }
+
+        private async Task WriteBatchesAndConfirmAsync(List<ShardedSubscriptionWorker.PublishedShardBatch> batches)
+        {
+            try
+            {
+                // Send to client
+                await WriteBatchToClientAndAckAsync(batches);
+
+                // let sharded subscription worker know that we sent the batch to the client and received an ack request from it
+                foreach (var batch in batches)
                 {
-                    result.Exceptions.Add(shard, e);
-                    result.ShardsToReconnect.Add(shard);
+                    batch.SendBatchToClientTcs.SetResult();
+                }
+            }
+            catch (Exception e)
+            {
+                // need to fail the shard subscription worker
+                foreach (var batch in batches)
+                {
+                    batch.SendBatchToClientTcs.SetException(e);
+                }
```

Comment:
I assume we need to add more tests to verify the correctness for those cases (regardless of the impl.)
Haven't actually checked that but in the previous impl. after a failure we would resend the same batch (but I think the client might get batches for other shard before getting the failed one)
In this impl. it depends.. haven't actually check it but I guess the client will get the not acked documents again in different batch (different docs combination) since for some shards we could ack and for other no.. 

In neither case I can see an actual issue.